File: leetcode/medium/701-insert-bst/iterative.cpp
/**
 * @file iterative.cpp
 * @author Vortexx2
 * @brief Problem 701 - Insert into a BST
 * @date 26-07-2021
 * 
 * Runtime - 84 ms
 * Memory Usage - 56.9 MB
 */

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  TreeNode *insertIntoBST(TreeNode *root, int val)
  {
    TreeNode *newNode = new TreeNode(val);

    if (!root)
    {
      root = newNode;
      return root;
    }

    TreeNode *prev = nullptr, *curr = root;

    while (curr)
    {
      prev = curr;
      if (curr->val < val)
      {
        curr = curr->right;
      }

      else
      {
        curr = curr->left;
      }
    }

    if (prev->val < val)
      prev->right = newNode;
    else
      prev->left = newNode;

    return root;
  }
};

File: leetcode/medium/701-insert-bst/iterative.py
"""
  @author Vortexx2
  Problem 701 - Insert into BST

  Runtime - 132 ms
  Memory Usage - 16.9 MB
"""
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def insertIntoBST(self, root: TreeNode, val: int) -> TreeNode:
    newNode = TreeNode(val)
    if not root:
      root = newNode
      return root

    curr, prev = root, None

    while curr:
      prev = curr
      if curr.val < val:
        curr = curr.right

      else:
        curr = curr.left

    if prev.val < val:
      prev.right = newNode

    else:
      prev.left = newNode

    return root
